### PR TITLE
Feature/add config option

### PR DESCRIPTION
**Pull request by murilo-cunha**
Add configuration file options via `pyproject.toml`, tests, docs.

Config is injected in context as defaults. That way we have the following args hierarchy (hight to low):

1. User-passed args
2. Config args
3. Default values

Config can be passed explicitly (`--config`) or implicitly (`pyproject.toml` file must be between common directory of targeted files and git repo root). The order of preference:

1. User-passed config file
2. If none, look for config file in path
3. If none is found, use default values

### CODE REVIEW

File: databooks/cli.py [hunk 1]
```diff
@@ -25,17 +27,54 @@
...
+    if config_path is not None:  # config may not be specified
+        with config_path.open("r") as f:
+            conf = (
+                tomli.load(f)
+                .get("tool", {})
+                .get("databooks", {})
+                .get(ctx.command.name, {})
+            )
+        # Merge configuration
+        ctx.default_map.update({k.replace("-", "_"): v for k, v in conf.items()})
```
Comment (frederikdesmedt) [inline databooks/cli.py:70]: What do you think about the way this is done here @murilo-cunha? My main worry is that we're linking app configuration to `click`. What do you think about having a `settings.py` or simply a `dict` that we manage ourselves? The responsibility of `typer` would then be to properly populate this `dict`/`Settings` object and to then call the right command. And this command would then be another function, e.g., `command.meta`, which would be responsible simply for executing the core functionality given certain parameters. Decoupling the core logic and the configuration management+command arg parsing+etc would provide us with a bunch of advantages, here's a non-exhaustive list:
- We can later change the command parsing tool if we want to without having to touch the core logic
- We can directly call the core functions (e.g., `command.meta`) in our tests, without having to go through the CLI parsing
- We have the option of supporting other front-ends in the future (e.g., a UI or webserver, assuming that we ever want to go that route of course)
- It simplifies the code of the functions themselves (`cli.meta` and `command.meta` in this case): one just deals with parameter management and parsing basically, the other just deals with "business logic"
- We can call our config resolution in our tests without actually having to assess the impact it would have on the core logic (e.g., we can test that `overwrite = true` is set in a certain use case by just asserting that the resulting `dict` has `overwrite = true`, rather than actually running it and verifying that it has overwritten).

File: databooks/cli.py [hunk 2]
```diff
@@ -45,9 +84,9 @@ def callback(  # noqa: D103
 callback.__doc__ = _DISTRIBUTION_METADATA["Summary"]
 
 
-@app.command()
+@app.command(add_help_option=False)
```
Comment (frederikdesmedt) [inline databooks/cli.py:87]: Why do we use our own help functionality? This arg seems to suggest that `typer` has something for this by default

File: databooks/logging.py [hunk 3]
```diff
@@ -1,12 +1,14 @@
...
 import logging
+import os
 
 from rich.logging import RichHandler
 
 
-def get_logger(name: str, level: str = "INFO") -> logging.Logger:
+def get_logger(name: str) -> logging.Logger:
     """Get logger with rich configuration."""
+    level = os.getenv("LOG_LEVEL", logging.INFO)
```
Comment (frederikdesmedt) [inline databooks/logging.py:11]: Don't we want to be able to overwrite this with a config file/cli arg? We can definitely do this in another PR as well, just mentioning it

File: docs/usage/configuration.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+metadata-head = false
+```
+
+## How can I use it?
+
+There are 2 ways to specify the configuration file: explicitly and implicitly. You can
+explicitly specify the `pyproject.toml` via the `--config` parameter. If none is specified,
+then `databooks` will look for a `pyproject.toml` in your project.
+
+`databooks` will look for the configuration file by first finding the common directory
```
Comment (frederikdesmedt) [inline docs/usage/configuration.md:56]: This is nice!

**Review by frederikdesmedt — COMMENTED** [review]
Looks nice!

For the CLI tests, looks like they're all integration tests now (testing everything from CLI parsing up to asserting the final output, e.g., the notebooks that were (not) overwritten). I think it would be good to be able to test more fine-grained units as well though. We could do this with one of the suggested changes I made for example. But let's see what we handle in this PR as well, as it's already quite big.

File: databooks/logging.py (hunk 3 above)
Comment (murilo-cunha) [reply databooks/logging.py:11]: Actually this came from [another PR](https://github.com/datarootsio/databooks/pull/4) but it got lost somewhere (probably when I was rebasing or something).
In the CLI, you can pass the `--verbose` argument that will pass it to downstream functions. When the argument is passed, we call the [`set_verbose(logger)`](https://github.com/datarootsio/databooks/pull/18/files/59bcacc8c1c751f546474d6ab78f919d9f440f5b#diff-45ceef528db0c2cc8bdd915a8031106ea55c828fd2d435c3bb47ad69774e3abfR21)
And since every CLI argument is configurable, then you can also configure it in the CLI 🚀 .

The thing I like about this approach is that we can still increase the amount of logs, cause when passing `--verbose` you only "reach" the functions that take that as an input. Whereas if you really want to debug stuff you can set all loggers everywhere. This is what I tried explaining [here](https://github.com/datarootsio/databooks/pull/18/files/59bcacc8c1c751f546474d6ab78f919d9f440f5b#diff-45ceef528db0c2cc8bdd915a8031106ea55c828fd2d435c3bb47ad69774e3abfR21)

File: databooks/cli.py (hunk 2 above)
Comment (murilo-cunha) [reply databooks/cli.py:87]: Yeah it does, but it's not an `eager` command. So if you type something like `databooks fix --help` it'd complain that you didn't pass `paths` (cause it'd try to get the config - which `is_eager` and depends on the paths).
If we don't reimplement it then only `databooks fix . --help` would work...

File: databooks/cli.py [hunk 5]
```diff
@@ -25,17 +27,54 @@
...
+def _version_callback(show_version: bool) -> None:
     """Return application version."""
-    if value:
+    if show_version:
         echo("databooks version: " + _DISTRIBUTION_METADATA["Version"])
         raise Exit()
 
 
+def _help_callback(ctx: Context, show_help: Optional[bool]) -> None:
+    """Reimplement `help` command to execute eagerly."""
```
[the remark this answers is not in the document]
Comment (murilo-cunha) [reply databooks/cli.py:38]: Actually we also [remove the default help command](https://github.com/datarootsio/databooks/pull/18/files/59bcacc8c1c751f546474d6ab78f919d9f440f5b#diff-0ed5aad2e701178980d9ece699ec53b77b8de9e871d6c5ac75d3f1c9821d1f59R87), and we reimplement it here. So there is no underlying command here anymore. Here we just print the `help` text and exit.

The only difference is that we make the `--help` argument `is_eager` so it won't mess up the execution

File: databooks/cli.py (hunk 1 above)
Comment (murilo-cunha) [reply databooks/cli.py:70]: If I understand you correctly, this is what I had started with, and I started `config.py` with that in mind. More specifically, I thought we could use [Pydantic models](https://pydantic-docs.helpmanual.io/usage/settings/) for this even (we can still add add the pydantic models later on to allow many file types). But I ran into some issues along the way.
So just to be clear, the my approach was:
1. Disregard the whole `is_eager` and `callback`
2. First thing that the `databooks.cli.{meta,fix}` does is to merge CLI arguments and config
3. Carry on with the rest of the program with the merged args
Also, the desired output is to have the parameter in following hierarchy (high priority to low):
1. CLI arguments (user passed arguments - ie: `--rm-outs`, etc.)
2. Config arguments (from `pyproject.toml`)
3. Default arguments

The issues that I had were:
1. Say we store the config in an object called `config` - we'd have to replace all instances of the `var` for `config.var`, which would make the code more polluted and require some hard coding (either specify different config objects for each command or specify a `config.command.var`
2. The first thing the command would do is find the config path (we need path that is passed in the CLI to find it), so the logic of injecting the config is with the logic of the command itself
3. **Most importantly,** if we do this in the command logic, then we cannot distinguish between what was defaults and what the user passed. So the config would always be overwritten by the default args (and break the order of hierarchy from before)

I tried different things and this was the cleanest solution. This is also the way that [`black` works](https://github.com/psf/black/blob/main/src/black/__init__.py#L110) (not that we have to do it the same, but when I was stuck and was comparing my ways out this seemed like the cleanest solution).

Final thoughts:
- This way I think we have the logic from the config almost entirely separate from the command (if you ignore the callbacks you'd be tricked into thinking that there is no config options)
- We can make the config logic very generic and we can reuse it
- This feels like using Click and it's features as it was intended. So we avoid reinventing the wheel.

Not sure I was missing something and if there is a more elegant way I'd be happy to refactor this!
Wdyt?

Comment (murilo-cunha) [reply databooks/cli.py:70]: Ah as for testing, I agree that maybe we should look into a way of testing the conf parsing more clearly. As I feel like this feature was made for this `ctx.default_map`, I'd imagine there is a clean way to test this.
I'll look more into it as well and add some tests if I find anything 😊

Comment (frederikdesmedt) [reply databooks/cli.py:70]: Ok, I can see what you're saying, it makes sense. Let's see for the testing indeed, if there is nothing like this out of the box for typer then I guess it could still make sense to separate the business logic and argument parsing overall, then it would be something like:
```
@app.command
def meta(the_typer_args_here...):
    cmd.meta(the_business_args...) # don't include stuff like `--config` here, because it doesn't matter for the meta logic
```
and then you can monkeypatch `cmd.meta` to see what business args the `meta` typer command came up with and sent to the `meta` command. And similarly you can test `cmd.meta` without having to essentially pass CLI options. A disadvantage is that there'll be a lot of overlap with the `cli.meta` and `cmd.meta` function arguments, so a lot of copy-pasting I can imagine... But let's see then how `typer` recommends to do it indeed. I'll resolve this thread for now as it's either way not super relevant for the PR

File: databooks/logging.py (hunk 3 above)
Comment (frederikdesmedt) [reply databooks/logging.py:11]: Ok cool, that makes sense!

File: databooks/cli.py (hunk 2 above)
Comment (frederikdesmedt) [reply databooks/cli.py:87]: I'm amazed that this is not the default 😅 Thanks for sharing the knowledge
